### PR TITLE
Allow scalar all gather

### PR DESCRIPTION
**Pull request by bfontain**
This change allows all gather to work for scalars. This is supported by the normal pytorch distributed API, but if you attempt to all gather scalars under xla, a low level error is emitted as the xla gather axis, 0, is not less than the rank, also 0, of the inputs.

This is fixed by reshaping the scalar tensors to shape (1,) and then reshaping the sliced outputs back to scalars.

This was tested manually on Cloud TPU.

### CODE REVIEW

File: torch_xla/distributed/xla_backend.py [hunk 1]
```diff
@@ -73,10 +73,14 @@ def allreduce(self, tensors, all_reduce_options):
...
     for input_tensor, output_tensors in zip(input_tensors, output_tensors_list):
+      is_scalar = (input_tensor.dim() == 0)
+      if is_scalar:
+        input_tensor = torch.reshape(input_tensor, (1,))
       result = xm.all_gather(input_tensor, groups=self._mesh, pin_layout=False)
       for i, slice in enumerate(torch.split(result, input_tensor.shape[0])):
         with torch.no_grad():
-          output_tensors[i].copy_(slice)
+          output_tensors[i].copy_(
+              slice if not is_scalar else torch.reshape(slice, ()))
```
Comment (JackCaoG) [inline torch_xla/distributed/xla_backend.py:84]: can you add a test case similar to  https://github.com/pytorch/xla/blob/d8c6a83ffd233c90d46f0544787dd3162edbae73/test/test_torch_distributed_xla_backend.py#L80-L88 ?

Comment (bfontain) [reply torch_xla/distributed/xla_backend.py:84]: Done! Validated that the test fails without this patch and passes with it.

**Review by JackCaoG — APPROVED** [review]
Thanks! I will merge it once CI passed. (If I forgot feel free to give me a ping)

**Comment by bfontain** [comment]
Ping on this PR, the failing tests seem unrelated to this change.

**Comment by JackCaoG** [comment]
sorry my bad.. should of merge it earlier. Do you mind resolve the merge conflict and rebase? Do you need this to backport to 2.2 release?

**Comment by bfontain** [comment]
Resolved the conflict and no need to backport.

**Comment by bfontain** [comment]
Thanks!
